include/arch/arm/aarch64/arm_mmu.h

```diff
@@ -33,25 +33,38 @@
  * current usage of attrs[31:3] is:
  * attrs[3] : Access Permissions
  * attrs[4] : Memory access from secure/ns state
- * attrs[5] : Execute Permissions
+ * attrs[5] : Execute Permissions privileged mode (PXN)
+ * attrs[6] : Execute Permissions unprivileged mode (UXN)
+ * attrs[7] : Mirror RO/RW permissions to EL0
  *
  */
 #define MT_PERM_SHIFT		3U
 #define MT_SEC_SHIFT		4U
-#define MT_EXECUTE_SHIFT	5U
+#define MT_P_EXECUTE_SHIFT	5U
+#define MT_U_EXECUTE_SHIFT	6U
+#define MT_RW_EL0h_SHIFT	7U
 
 #define MT_RO			(0U << MT_PERM_SHIFT)
 #define MT_RW			(1U << MT_PERM_SHIFT)
 
+#define MT_RW_EL0h		(1U << MT_RW_EL0h_SHIFT)
+#define MT_RW_ELh		(0U << MT_RW_EL0h_SHIFT)
+
 #define MT_SECURE		(0U << MT_SEC_SHIFT)
 #define MT_NS			(1U << MT_SEC_SHIFT)
 
-#define MT_EXECUTE		(0U << MT_EXECUTE_SHIFT)
-#define MT_EXECUTE_NEVER	(1U << MT_EXECUTE_SHIFT)
+#define MT_P_EXECUTE		(0U << MT_P_EXECUTE_SHIFT)
+#define MT_P_EXECUTE_NEVER	(1U << MT_P_EXECUTE_SHIFT)
+
+#define MT_U_EXECUTE		(0U << MT_U_EXECUTE_SHIFT)
+#define MT_U_EXECUTE_NEVER	(1U << MT_U_EXECUTE_SHIFT)
 
-/* Some compound attributes for most common usages */
-#define MT_CODE			(MT_NORMAL | MT_RO | MT_EXECUTE)
-#define MT_RODATA		(MT_NORMAL | MT_RO | MT_EXECUTE_NEVER)
+#define MT_P_RW_U_RW		(MT_RW | MT_RW_EL0h | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RW_U_NA		(MT_RW | MT_RW_ELh  | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RO_U_RO		(MT_RO | MT_RW_EL0h | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RO_U_NA		(MT_RO | MT_RW_ELh  | MT_P_EXECUTE_NEVER | MT_U_EXECUTE_NEVER)
+#define MT_P_RX_U_RX		(MT_RO | MT_RW_EL0h | MT_P_EXECUTE | MT_U_EXECUTE)
```

Comment:
I assume you intend this type to map Application code.
This can create security hole. You don't want some thing from such section to be able to run at higher privilege.
MT_P_EXECUTE_NEVER |  MT_U_EXECUTE is better.

Reply:
I understand your point but this list is just used to define the most common combinations, we don't have to necessarily use it. I can add a `MT_P_R_U_RX`.

BTW this is also mapping what's defined on other architectures as well: https://github.com/zephyrproject-rtos/zephyr/blob/master/include/arch/arc/arch.h#L218, https://github.com/zephyrproject-rtos/zephyr/blob/master/include/arch/arm/aarch32/cortex_m/mpu/arm_mpu_v7m.h#L157, https://github.com/zephyrproject-rtos/zephyr/blob/master/include/arch/x86/mmustructs.h#L144

Reply:
Added `MT_P_RO_U_RX`